tools/analysis/src/futures.rs

```diff
@@ -0,0 +1,72 @@
+use core::{
+    future::Future,
+    mem::swap,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use futures::{future::FutureExt, pin_mut};
+
+pub struct JoinedFuture<F: Future> {
+    futures: Vec<(Pin<Box<F>>, Poll<F::Output>)>,
+}
+
+impl<F: Future> JoinedFuture<F> {
+    pub fn new<I>(futures: I) -> Self
+    where
+        I: IntoIterator<Item = F>,
+        F: Future,
+    {
+        JoinedFuture {
+            futures: futures
+                .into_iter()
+                .map(|fut| (Box::pin(fut), Poll::Pending))
+                .collect(),
+        }
+    }
+}
+
+impl<F: Future> Future for JoinedFuture<F>
+where
+    F::Output: Unpin,
+{
+    type Output = Vec<F::Output>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut finished = true;
+
+        for (fut, res) in self.futures.iter_mut() {
+            match res {
+                Poll::Pending => {
+                    let _fut = fut;
+                    pin_mut!(_fut);
```

Comment:
Can we just get a Pin<> reference out of the vector, since we already have a vector of pin boxes ?